iree/turbine/kernel/wave/index_sequence_analysis.py

```diff
@@ -668,3 +676,89 @@ def apply_offset(node: fx.Node):
         return False
 
     trace.walk(apply_offset)
+
+
+def create_broadcast(
+    binary_op: BinaryPyOp,
+    to_broadcast: CustomOp,
+    broadcast_dim: IndexSymbol,
+    broadcast_size: int,
+    target_node: CustomOp,
+):
+    """
+    Create a broadcast node for the given binary operator.
+    """
+    with binary_op.graph.inserting_before(binary_op.fx_node):
+        broadcasted = Broadcast(to_broadcast.fx_node, target_node.type).add_to_graph(
+            binary_op.graph
+        )
+        custom = get_custom(broadcasted)
+        custom.vector_shapes = to_broadcast.vector_shapes
+        custom.index = deepcopy(target_node.index)
+        custom.index[broadcast_dim].size = broadcast_size
+        broadcast_idx = list(binary_op.node_args.values()).index(to_broadcast)
+        binary_op.update_arg(broadcast_idx, custom.fx_node)
+
+
+def resolve_thread_shapes(trace: CapturedTrace, constraints: list[Constraint]):
+    """
+    This function walks through all the binary operators in the graph and
+    if there is a discrepancy between the thread shapes of the operators
+    along the same dimension it resolves the discrepancy.
+
+    Currently, the only mismatches that can be resolved are when one of
+    the shapes is 1 and the other is > 1.
+    """
+    binary_ops = trace.walk(lambda node: isinstance(get_custom(node), BinaryPyOp))
+    for binary_op in binary_ops:
+        custom = get_custom(binary_op)
+        # Get the largest dim and shape from the lhs and rhs.
+        lhs = get_custom(custom.lhs)
+        rhs = get_custom(custom.rhs)
+        lhs_dim, lhs_size = get_largest_index_and_size(lhs.index)
+        rhs_dim, rhs_size = get_largest_index_and_size(rhs.index)
+        if lhs_size > rhs_size:
+            to_broadcast = rhs
+            broadcast_dim = lhs_dim
+            broadcast_size = lhs_size
+            target = lhs
+        else:
+            to_broadcast = lhs
+            broadcast_dim = rhs_dim
+            broadcast_size = rhs_size
+            target = rhs
+        # If they are equal we are done.
+        if lhs_dim == rhs_dim and lhs_size == rhs_size:
+            continue
+        # If all are unit dims, there is nothing to do.
+        if lhs_size == 1 and rhs_size == 1:
+            continue
+        if lhs_dim != rhs_dim:
+            # If the dimensions don't agree, we can still do this broadcast only if
+            # this has a reduce op in its backward slice along the broadcasting dimension,
```

Comment:
Can w get an IR/visual example for this :)

Reply:
Sure will add it into the comments.

Reply:
okay I simplified this a bit. Can you take another look?